File: 0349_IntersectionOfTwoArrays/Jason/349.go
package main

func main() {

}

func Intersection(nums1 []int, nums2 []int) []int {
	// Version 1
	result := []int{}
	tmp := make(map[int]bool)
	nums1Set := []int{}

	for _, num := range nums1 {
		if _, ok := tmp[num]; !ok {
			tmp[num] = false
			nums1Set = append(nums1Set, num)
		}
	}

	for _, num := range nums2 {
		if _, ok := tmp[num]; ok {
			tmp[num] = true
		}
	}

	for _, num := range nums1Set {
		if value, _ := tmp[num]; value {
			result = append(result, num)
		}
	}

	return result
}

File: 0349_IntersectionOfTwoArrays/Jason/349.cpp
#include <iostream>
#include <vector>
#include <set>
#include <map>

using namespace std;

int main() {
    
}

vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
    // Version 2 - map
    map<int, bool> m;
    vector<int> result;

    for (vector<int>::iterator it = nums1.begin(); it != nums1.end(); ++it) {
        m[*it] = false;
    }

    for (vector<int>::iterator it = nums2.begin(); it != nums2.end(); ++it) {
        if (m.find(*it) != m.end())
            m[*it] = true;
    }

    for (map<int, bool>::iterator it = m.begin(); it != m.end(); ++it) {
        if (it->second)
            result.push_back(it->first);
    }

    return result;

    // Version 1 - Set
    // set<int> s1, s2;
    // vector<int> result;
    // for (int i = 0; i < nums1.size(); ++i) {
    //     s1.insert(nums1[i]);
    // }

    // for (int i = 0; i < nums2.size(); ++i) {
    //     if (s1.find(nums2[i]))
    //         s2.insert(nums2[i]);
    // }

    // for (set<int>::iterator it = s2.begin(); it != s2.end(); ++it) {
    //     result.push_back(*it);
    // }

    // return result;
}